src/Lean/MonadEnv.lean

```diff
@@ -152,26 +152,26 @@ private def checkUnsupported [Monad m] [MonadEnv m] [MonadError m] (decl : Decla
     | some (Expr.const declName ..) => throwError "code generator does not support recursor '{declName}' yet, consider using 'match ... with' and/or structural recursion"
     | _ => pure ()
 
-def compileDecl [Monad m] [MonadEnv m] [MonadError m] [MonadOptions m] (decl : Declaration) : m Unit := do
-  match (← getEnv).compileDecl (← getOptions) decl with
+def compileDecl [Monad m] [MonadEnv m] [MonadError m] [MonadOptions m] (maxHeartbeats: Nat) (decl : Declaration) : m Unit := do
+  match (← getEnv).compileDecl maxHeartbeats (← getOptions) decl with
   | Except.ok env   => setEnv env
   | Except.error (KernelException.other msg) =>
     checkUnsupported decl -- Generate nicer error message for unsupported recursors and axioms
     throwError msg
   | Except.error ex =>
     throwKernelException ex
 
-def compileDecls [Monad m] [MonadEnv m] [MonadError m] [MonadOptions m] (decls : List Name) : m Unit := do
-  match (← getEnv).compileDecls (← getOptions) decls with
+def compileDecls [Monad m] [MonadEnv m] [MonadError m] [MonadOptions m] (maxHeartbeats: Nat) (decls : List Name) : m Unit := do
+  match (← getEnv).compileDecls maxHeartbeats (← getOptions) decls with
   | Except.ok env   => setEnv env
   | Except.error (KernelException.other msg) =>
     throwError msg
   | Except.error ex =>
     throwKernelException ex
 
-def addAndCompile [Monad m] [MonadEnv m] [MonadError m] [MonadOptions m] [MonadLog m] [AddMessageContext m] (decl : Declaration) : m Unit := do
-  addDecl decl;
-  compileDecl decl
+def addAndCompile [Monad m] [MonadEnv m] [MonadError m] [MonadOptions m] [MonadLog m] [AddMessageContext m] (maxHeartbeats: Nat) (decl : Declaration) : m Unit := do
```

Comment:
```suggestion
def addAndCompile (decl : Declaration) : CoreM Unit := do
```
I think we should avoid the extra argument here.  (And adding a `MonadHeartbeat` class feels overkill here, this isn't even a combinator.)